core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/domain/variable/listener/support/AbstractNotifiable.java

```diff
@@ -32,37 +32,31 @@ static <Solution_> EntityNotifiable<Solution_> buildNotifiable(
             return new ListVariableListenerNotifiable<>(
                     scoreDirector,
                     ((ListVariableListener<Solution_, Object>) variableListener),
-                    globalOrder);
+                    new ArrayDeque<>(), globalOrder);
         } else {
+            VariableListener<Solution_, Object> basicVariableListener = (VariableListener<Solution_, Object>) variableListener;
             return new VariableListenerNotifiable<>(
                     scoreDirector,
-                    ((VariableListener<Solution_, Object>) variableListener),
+                    basicVariableListener,
+                    basicVariableListener.requiresUniqueEntityEvents()
+                            ? new ListBasedScalingOrderedSet<>()
```

Comment:
Side note: ever since playing with this collection in my CS performance experiments, I am not sure that it actually brings performance benefits. One of those days, it needs to be benchmarked and possibly removed.

Reply:
I agree, avoid using ListBasedScalingOrderedSet()
For CS, it might become unused with the TupleList refactor

Reply:
https://issues.redhat.com/browse/PLANNER-2809